### PR TITLE
[Nim] Add basics Concept Exercise

### PR DESCRIPTION
**Pull request by iHiD**
Added by @iHiD on behalf of @ynfle as our auto-upload failed 🙂

### CODE REVIEW

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+In this exercise you're going to write some code to help you cook a brilliant lasagna from your favorite cooking book.
+
+You have four tasks, all related to the time spent cooking the lasagna.
+
+1. Define the expected oven time in minutes
+   Define the expectedMinutesInOven constant that returns how many minutes the lasagna should be in the oven. It must be exported. According to the cooking book, the expected oven time in minutes is 40:
+
+expectedMinutesInOven
+// => 40
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:13]:
````suggestion
```nim
expectedMinutesInOven
// => 40
```
````

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+2. Calculate the remaining oven time in minutes
+   Implement the remainingMinutesInOven function that takes the actual minutes the lasagna has been in the oven as a parameter and returns how many minutes the lasagna still has to remain in the oven, based on the expected oven time in minutes from the previous task.
+
+remainingMinutesInOven(30)
+// => 10
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:19]:
````suggestion
```nim
remainingMinutesInOven(30)
// => 10
```
````

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+3. Calculate the preparation time in minutes
+   Implement the preparationTimeInMinutes function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
+
+preparationTimeInMinutes(2)
+// => 4
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:25]:
````suggestion
```nim
preparationTimeInMinutes(2)
// => 4
```
````

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+4. Calculate the total working time in minutes
+   Implement the totalTimeInMinutes function that takes two parameters: the numberOfLayers parameter is the number of layers you added to the lasagna, and the actualMinutesInOven parameter is the number of minutes the lasagna has been in the oven. The function should return how many minutes in total you've worked on cooking the lasagna, which is the sum of the preparation time in minutes, and the time in minutes the lasagna has spent in the oven at the moment.
+
+totalTimeInMinutes(3, 20)
+// => 26
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:31]:
````suggestion
```nim
totalTimeInMinutes(3, 20)
// => 26
```
````

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+// => 10
+
+3. Calculate the preparation time in minutes
+   Implement the preparationTimeInMinutes function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
+
+preparationTimeInMinutes(2)
+// => 4
+
+4. Calculate the total working time in minutes
+   Implement the totalTimeInMinutes function that takes two parameters: the numberOfLayers parameter is the number of layers you added to the lasagna, and the actualMinutesInOven parameter is the number of minutes the lasagna has been in the oven. The function should return how many minutes in total you've worked on cooking the lasagna, which is the sum of the preparation time in minutes, and the time in minutes the lasagna has spent in the oven at the moment.
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:28]:
```suggestion
   Implement the `totalTimeInMinutes` function that takes two parameters: the `numberOfLayers` parameter is the number of layers you added to the lasagna, and the `actualMinutesInOven` parameter is the number of minutes the lasagna has been in the oven. The function should return how many minutes in total you've worked on cooking the lasagna, which is the sum of the preparation time in minutes, and the time in minutes the lasagna has spent in the oven at the moment.
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+// => 40
+
+2. Calculate the remaining oven time in minutes
+   Implement the remainingMinutesInOven function that takes the actual minutes the lasagna has been in the oven as a parameter and returns how many minutes the lasagna still has to remain in the oven, based on the expected oven time in minutes from the previous task.
+
+remainingMinutesInOven(30)
+// => 10
+
+3. Calculate the preparation time in minutes
+   Implement the preparationTimeInMinutes function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:22]:
```suggestion
   Implement the `preparationTimeInMinutes` function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+You have four tasks, all related to the time spent cooking the lasagna.
+
+1. Define the expected oven time in minutes
+   Define the expectedMinutesInOven constant that returns how many minutes the lasagna should be in the oven. It must be exported. According to the cooking book, the expected oven time in minutes is 40:
+
+expectedMinutesInOven
+// => 40
+
+2. Calculate the remaining oven time in minutes
+   Implement the remainingMinutesInOven function that takes the actual minutes the lasagna has been in the oven as a parameter and returns how many minutes the lasagna still has to remain in the oven, based on the expected oven time in minutes from the previous task.
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:16]:
```suggestion
   Implement the `remainingMinutesInOven` function that takes the actual minutes the lasagna has been in the oven as a parameter and returns how many minutes the lasagna still has to remain in the oven, based on the expected oven time in minutes from the previous task.
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 8]
```diff
@@ -0,0 +1,31 @@
+You're going to write some code to help you cook a brilliant lasagna from your favorite cookbook.
+
+You have five tasks, all related to cooking your recipe.
+
+In this exercise you're going to write some code to help you cook a brilliant lasagna from your favorite cooking book.
+
+You have four tasks, all related to the time spent cooking the lasagna.
+
+1. Define the expected oven time in minutes
+   Define the expectedMinutesInOven constant that returns how many minutes the lasagna should be in the oven. It must be exported. According to the cooking book, the expected oven time in minutes is 40:
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:10]:
```suggestion
   Define the `expectedMinutesInOven` constant that returns how many minutes the lasagna should be in the oven. It must be exported. According to the cooking book, the expected oven time in minutes is 40:
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,31 @@
+You're going to write some code to help you cook a brilliant lasagna from your favorite cookbook.
+
+You have five tasks, all related to cooking your recipe.
+
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:4]: This looks like some duplication

```suggestion
```

File: languages/nim/exercises/concept/basics/.meta/example.nim [hunk 10]
```diff
@@ -0,0 +1,19 @@
+const expectedMinutesInOven* = 40
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.meta/example.nim:1]: Is the `*` the "export identifiers" concept listed in the "Learning concepts"?

File: languages/nim/exercises/concept/basics/.meta/design.md [hunk 11]
```diff
@@ -0,0 +1,22 @@
+## Learning objectives
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.meta/design.md:1]: I think this list should also include knowing how to define comments (as the stub file will have them)

File: languages/nim/exercises/concept/basics/.meta/design.md [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+- Know `const` definition
+- Know about type inference
+- Know about `proc` parameters
+- Know the different way of returning values from a `proc` (implicit, `return`, `result`)
+- Know how to export identifiers
+- Know how to print values for debugging (`echo`, `debugEcho`)
+
+## Out of scope
+
+- `func` (I think we'll do another exercise on this so don't what to put here)
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.meta/design.md:13]: What is the difference between "Know about `proc` definition" and a `func`?

> I think we'll do another exercise on this so don't what to put here

What do you plan on doing in that other exercise? You'll probably be doing some advanced function concepts, right? If so, you can list those advanced concepts here.

File: languages/nim/exercises/concept/basics/.meta/design.md [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+- `var`, `let`, `var` assignment
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.meta/design.md:14]:
```suggestion
- `var`, `let` assignment
```

Comment (ErikSchierboom) [reply languages/nim/exercises/concept/basics/.meta/design.md:14]: One thing to consider is that any subsequent exercise won't be able to use `var` or `let`, unless they depend on another exercise in which these concepts are introduced. This is totally valid of course, but is something to consider.

**Review by ErikSchierboom — APPROVED** [review]
Great work! I've left some comments.

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+In this exercise you're going to write some code to help you cook a brilliant lasagna from your favorite cooking book.
+
+You have four tasks, all related to the time spent cooking the lasagna.
+
+1. Define the expected oven time in minutes
```
Comment (iHiD) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:9]: @ErikSchierboom This headings should be actual headings, right?

Comment (ErikSchierboom) [reply languages/nim/exercises/concept/basics/.docs/instructions.md:9]: Yep!

Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:9]:
```suggestion
## 1. Define the expected oven time in minutes
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+   Define the expectedMinutesInOven constant that returns how many minutes the lasagna should be in the oven. It must be exported. According to the cooking book, the expected oven time in minutes is 40:
+
+expectedMinutesInOven
+// => 40
+
+2. Calculate the remaining oven time in minutes
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:15]:
```suggestion
## 2. Calculate the remaining oven time in minutes
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+   Implement the remainingMinutesInOven function that takes the actual minutes the lasagna has been in the oven as a parameter and returns how many minutes the lasagna still has to remain in the oven, based on the expected oven time in minutes from the previous task.
+
+remainingMinutesInOven(30)
+// => 10
+
+3. Calculate the preparation time in minutes
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:21]:
```suggestion
## 3. Calculate the preparation time in minutes
```

File: languages/nim/exercises/concept/basics/.docs/instructions.md [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+   Implement the preparationTimeInMinutes function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
+
+preparationTimeInMinutes(2)
+// => 4
+
+4. Calculate the total working time in minutes
```
Comment (ErikSchierboom) [inline languages/nim/exercises/concept/basics/.docs/instructions.md:27]:
```suggestion
## 4. Calculate the total working time in minutes
```

**Review by ErikSchierboom — APPROVED** [review]

File: languages/nim/exercises/concept/basics/.meta/example.nim (hunk 10 above)
Comment (ynfle) [reply languages/nim/exercises/concept/basics/.meta/example.nim:1]: Yes

File: languages/nim/exercises/concept/basics/.meta/design.md (hunk 12 above)
Comment (ynfle) [reply languages/nim/exercises/concept/basics/.meta/design.md:13]: I have no plans.
`func` is closer to the mathematical definition, ie. can't have side effects like modifying global variables, writing to files etc.

**Comment by ynfle** [comment]
I can't currently push to this PR. Should I close it and then submit a new one? Could someone with creds for this bot account just go in and check the "allow maintainers to push" checkbox?

File: languages/nim/exercises/concept/basics/.meta/design.md (hunk 13 above)
Comment (ynfle) [reply languages/nim/exercises/concept/basics/.meta/design.md:14]: It's not so important, that could even be the next exercise.
`let` is immutable and therefore can not be assigned to.
`var` is mutable and therefore there is the initialization of the variable and the assignment of the variable afterward.

**Comment by ynfle** [comment]
Closed in favour of #2214

**Comment by iHiD** [comment]
@ynfle I believe that by default that is the case. However as this is my PR, not the bots, it doesn't.

File: languages/nim/exercises/concept/basics/.meta/design.md (hunk 13 above)
Comment (ErikSchierboom) [reply languages/nim/exercises/concept/basics/.meta/design.md:14]: > It's not so important, that could even be the next exercise.

Sounds good

**Comment by ynfle** [comment]
Oh ok.
I didn't try pushing because it originated from `exercism-external-bot`.
I should've realized that it would've been different.

I can reopen this PR if you'd like

**Comment by iHiD** [comment]
New PR is totally fine :)

**Comment by ynfle** [comment]
Ok great!!
If you have time, I'd love your review on it also. Specifically on [this comment](https://github.com/exercism/v3/pull/2214#issuecomment-681150308) regarding the granularity, depth and breadth of the concept exercise.

**Comment by iHiD** [comment]
Sweet. I'll try and find some time tomorrow :)

**Comment by ynfle** [comment]
> Sweet. I'll try and find some time tomorrow :)

🙏 
No pressure!

**Pull request closed without merging**